'string signature' under python 2

When running ``dbusmock`` under python 2 unicode strings passed to a mocked method cause an error when they contain characters that can not be encoded using ascii.

According to the [dbus docs](https://dbus.freedesktop.org/doc/dbus-python/api/dbus.String-class.html) dbus strings that can be specified in a signature default to unicode, if the ``utf8_strings`` kwarg has not been set to ``True`` in the method decorator.

When passing such a unicode string containing non ascii chars to a mocked method with ``in_sig='s'`` the following error occurs:

```
E       DBusException: org.freedesktop.DBus.Python.UnicodeEncodeError: Traceback (most recent call last):
E         File "/usr/lib/python2.7/dist-packages/dbus/service.py", line 707, in _message_cb
E           retval = candidate_method(self, *args, **keywords)
E         File "/home/elbenfreund/.virtualenvs/hamster-dbus/local/lib/python2.7/site-packages/dbusmock/mockobject.py", line 330, in <lambda>
E           self, interface, name, in_sig, *args, **kwargs)
E         File "/home/elbenfreund/.virtualenvs/hamster-dbus/local/lib/python2.7/site-packages/dbusmock/mockobject.py", line 551, in mock_method
E           self.log(dbus_method + self.format_args(args))
E         File "/home/elbenfreund/.virtualenvs/hamster-dbus/local/lib/python2.7/site-packages/dbusmock/mockobject.py", line 599, in format_args
E           s += format_arg(a)
E         File "/home/elbenfreund/.virtualenvs/hamster-dbus/local/lib/python2.7/site-packages/dbusmock/mockobject.py", line 578, in format_arg
E           return '"' + str(a) + '"'
E       UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-9: ordinal not in range(128)
```

At a first glance this seems to be due to the use of ``str`` in L578 which means something different between py2 and py3. Maybe using ``six.text_type`` or something similar can avoid a forced type conversion here?

Please let me know if I can provide any further information to help sorting this issue
Eric

--- a/dbusmock/mockobject.py
+++ b/dbusmock/mockobject.py
@@ -90,7 +90,7 @@ class DBusMockObject(dbus.service.Object):
         self._template_parameters = None
 
         if logfile:
-            self.logfile = open(logfile, 'w')
+            self.logfile = open(logfile, 'wb')
         else:
             self.logfile = None
         self.is_logfile_owner = True
@@ -589,8 +589,10 @@ class DBusMockObject(dbus.service.Object):
                 return str(int(a))
             if isinstance(a, int) or isinstance(a, long):
                 return str(a)
-            if isinstance(a, str) or isinstance(a, unicode):
+            if isinstance(a, str):
                 return '"' + str(a) + '"'
+            if isinstance(a, unicode):  # Python 2 only
+                return '"' + repr(a.encode('UTF-8'))[1:-1] + '"'
             if isinstance(a, list):
                 return '[' + ', '.join([format_arg(x) for x in a]) + ']'
             if isinstance(a, dict):
@@ -623,12 +625,11 @@ class DBusMockObject(dbus.service.Object):
         otherwise it goes to stdout.
         '''
         if self.logfile:
-            fd = self.logfile
+            fd = self.logfile.fileno()
         else:
-            fd = sys.stdout
+            fd = sys.stdout.fileno()
 
-        fd.write('%.3f %s\n' % (time.time(), msg))
-        fd.flush()
+        os.write(fd, ('%.3f %s\n' % (time.time(), msg)).encode('UTF-8'))
 
     @dbus.service.method(dbus.INTROSPECTABLE_IFACE,
                          in_signature='',

--- a/tests/test_api.py
+++ b/tests/test_api.py
@@ -27,6 +27,9 @@ from gi.repository import GLib
 
 dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)
 
+# "a <heart> b" in py2/3 compatible unicode
+UNICODE = b'a\xe2\x99\xa5b'.decode('UTF-8')
+
 
 class TestAPI(dbusmock.DBusTestCase):
     '''Test dbus-mock API'''
@@ -78,6 +81,12 @@ class TestAPI(dbusmock.DBusTestCase):
         self.dbus_mock.AddMethod('', 'Do', 's', 's', 'ret = args[0]')
         self.assertEqual(self.dbus_test.Do('Hello'), 'Hello')
 
+    def test_unicode_str(self):
+        '''unicode string roundtrip'''
+
+        self.dbus_mock.AddMethod('', 'Do', 's', 's', 'ret = args[0] * 2')
+        self.assertEqual(self.dbus_test.Do(UNICODE), dbus.String(UNICODE * 2))
+
     def test_twoarg_ret(self):
         '''two arguments, code for return value'''
 
@@ -91,19 +100,24 @@ class TestAPI(dbusmock.DBusTestCase):
     def test_array_arg(self):
         '''array argument'''
 
-        self.dbus_mock.AddMethod('', 'Do', 'iaou', '',
-                                 '''assert len(args) == 3
+        self.dbus_mock.AddMethod('', 'Do', 'iaous', '',
+                                 '''assert len(args) == 4
 assert args[0] == -1;
 assert args[1] == ['/foo']
 assert type(args[1]) == dbus.Array
 assert type(args[1][0]) == dbus.ObjectPath
 assert args[2] == 5
-''')
-        self.assertEqual(self.dbus_test.Do(-1, ['/foo'], 5), None)
+assert args[3] == %s
+''' % repr(UNICODE))
+        self.assertEqual(self.dbus_test.Do(-1, ['/foo'], 5, UNICODE), None)
 
         # check that it's logged correctly
-        with open(self.mock_log.name) as f:
-            self.assertRegex(f.read(), r'^[0-9.]+ Do -1 \["/foo"\] 5$')
+        with open(self.mock_log.name, "rb") as f:
+            log = f.read()
+            if sys.version_info[0] >= 3:
+                self.assertRegex(log, b'^[0-9.]+ Do -1 \["/foo"\] 5 "a\\xe2\\x99\\xa5b"$')
+            else:
+                self.assertRegex(log, r'^[0-9.]+ Do -1 \["/foo"\] 5 "a\\xe2\\x99\\xa5b"$')
 
     def test_dict_arg(self):
         '''dictionary argument'''

